Issue1382 2 add path_only, name_only, data_only.  Improved documentation.

petersilva opened the pull request:
close  #1382 

* Does what was discussed, including some unit tests (more than before) 
* add path_only, name_only, rename data to data_only (but accept data as synonym.)
* documentation updated and (hopefully) clarified to give more concise guidance in the reference man page, but more robust motivation in the Explanation *DuplicateSuppression.rst* guide.  Also added a more full discussion about the Identity field in that same document.
*  in unit_tests message formats corrected to they better exercise the nodupe algorithm. Before this work,
    messages had *sum* fields  instead of *integrity*) so the "key" stuff in the code was never exercised.


path_only is what the original issue was complaining about.  it was expecting *path* to mean *path_only*.

reidsunderland committed:
nodupe tweaks, still needs more thorough documentation

reidsunderland committed:
comment making it more obvious

petersilva committed:
Correcting messages in unit tests so they are closer to ok.

petersilva committed:
nodupe unit tests and mock data updated to reality.

the old unit tests weren't exercising the duplicate suppression
logic because incorrect fields were present in the messages.
in place of *sum* field, need *identity* and format is different.
Also *parts* is replaced by *size*, *headers* is not a field...
too many to list.  Anyways, the test messages now have relevant
data.

petersilva committed:
Merge branch 'issue1382_unit_tests' into issue1382_2

petersilva committed:
initial nodupe_method path unit test

petersilva committed:
Merge branch 'issue1382' into issue1382_2

petersilva committed:
completing work on #1382 nodupe path work.

* updated unit tests to match @reid's nodupe changes.
* renamed nodupe/data plugin to data_only for consistency with path, and name.
* updated English documentation to reflect changes.
* renamed unit tests to reflect *only* naming.

petersilva committed:
French doc changes to match English ones in previous commit

andreleblanc11 commented:
The code changes look good to me. I like what's added to the documentation, it makes it really clear how the duplicate suppression works now.

petersilva commented:
If you're happy... please approve the PR...

reidsunderland commented:
I'm reviewing too, it looks awesome, thanks @petersilva !

So far I noticed a debug log line that I'll make a commit to remove

reidsunderland committed:
remove debug print:

reidsunderland commented on sarracenia/flowcb/nodupe/__init__.py:
```diff
@@ -28,24 +28,42 @@ class NoDupe(FlowCB):
+            # if file size is known, use it
             if 'size' in msg:
                 key = f"{msg['relPath']},{t},{msg['size']}"
             else:
```
When I first started looking at this, @andreleblanc11 was concerned that when identity is not available in the message `name` wouldn't work properly if the key is derived from `relPath` here. I think he's correct.

Should we check if `m['nodupe_override']['path']` is set, and if it is, use it here instead of `msg['relPath']`? 

(My code block selection didn't quite work). This is the section I'm wondering about:

https://github.com/MetPX/sarracenia/blob/8f8e90d9f18fc71b615e988b6900a486daf2db2a/sarracenia/flowcb/nodupe/__init__.py#L67-L70

reidsunderland replied:
This is what I think we might need?

```python3
 if 'nodupe_override' in msg and 'path' in msg['nodupe_override']:
     path = msg['nodupe_override']['path']
 else:
     path = msg['relPath']

 if 'size' in msg: 
     key = f"{path},{t},{msg['size']}" 
 else: 
     key = f"{path},{t}" 
```

reidsunderland reviewed (commented):
I added one note

petersilva committed:
use path_override in key calculation when identity missing.

Suggestion from André and Reid... probably right... I am not sure
myself... makes some sense.
